File: Lista4/exec4/loja.py
from estoque import *
from livro import *
from CD import *
from DVD import *

class loja:

    #  * cadastro de produto
    def __init__(self) -> None:
        self.itens = ['' for i in range(0, 1000)]
        self.modo = -1 
        self.qte = 0 
        self.num_itens = 0
        self.nome = ""     

    def cadastrarProduto(self) -> None:
        
        print('\033[32m' + '\nAdicionar Item\n'.upper() + '\033[0;0m')

        print("1) Inserir Livro")
        print("2) Inserir CD")
        print("3) Inserir DVD\n")

        try: 
            self.modo = int(input('> ')) 
        except Exception as e:
            print(f"\nErro: {e}")
            return

        
        try:
            cbarras = int(input('Código de barras para este produto: '))
            nome = input('Insira nome do produto: ')
            try:
                self.qte = int(input("Insira uma qte inicial para o estoque se quiser: "))
                if self.qte < 0: 
                    print("\nErro: A quantidade inicial minima é 0")
                    self.qte = 0
            except Exception as e: 
                print(f'Erro: {e}')
                self.qte = 0 

            if self.modo == 1:
                self.itens[self.num_itens] = livro(nome, cbarras, self.qte) 
            elif self.modo == 2: 
                self.itens[self.num_itens] = CD(nome, cbarras, self.qte) 
            elif 3: 
                self.itens[self.num_itens] = DVD(nome, cbarras, self.qte) 
            else:
                print("\nErro: Houve um problema ao ler o modo")

            if (self.ctProduto_codigo(cbarras) != -1) or (self.ctProduto_codigo(nome) != -1): 
                print("Erro: produto já cadastrado\n") 

        except Exception as e: 
            print(f"\nErro: {e}") 
            return 

        print(self.itens[self.num_itens])
        self.num_itens += 1
    
    #  * Adicionar mais estoque a um item ja existente
    def adicionarProduto(self):
        flag = ""
        self.qte = 0

        print('\033[32m' + "\nAdicionar Estoque\n".upper() + '\033[0;0m')

        pos = self.ctIndice()
        if pos == -1: 
            print("Erro \nItem não encontrado") 
            return 

        try:
            print("Item "+ str((self.itens[pos]).getNome()) + "encontrado")
            print("Insira a quantidade que deseja acrescentar: ")
            self.qte = input()

            if self.qte <= 0: 
                print("\nErro: Quantia inválida") 
                return 

        except Exception as e: 
            print(f"\nErro: {e}") 
            return 

        self.itens[pos].adicionar(self.qte)
        print("\n Quantia adicionada com sucesso ")
        print("Quantidade atual: "+ str(self.itens[pos].getself.Qte()))

    
    #  * Encontrar produtos através do código de barras ou do nomne
    #  * @param chave
    #  * @return
    def ctProduto_codigo(self, chave):
        if type(chave) == 'int':
            for item in self.itens:
                if item != '':
                    if self.num_iten == item.getNum(): 
                        return i
        else:
            for item in self.itens:
                if item != '':
                    if self.nome ==  item.getNome(): 
                        return i 

        return -1


    #  * Le dados de busca do usuário e retorna a posição do produto
    #  * @return
    def ctIndice(self):
        flag = ""
        self.qte = 0
        self.pos = -1

        #try:
        flag = input('Insira o nome ou código de barras do item que deseja modificar: ')
        try:
            self.num_itens = int(flag)                
            pos = self.ctProduto_codigo(self.cbarras)
            if pos == -1: 
                pos = self.ctProduto_codigo(flag)  
        except Exception as e:
            pos = self.ctProduto_codigo(flag)
            
        #except Exception as e: 
        #    print(f"\nErro:{e}") 
        #    return -2 

        return pos

    #  * buscar e imprimir certo produto
    def getProduto(self):
        print('\033[32m' + "\nBuscar Item\n".upper() + '\033[0;0m')
        pos = self.ctIndice()
        if pos == -1: 
            print("\nErro: Item não encontrado") 
            return 
        print("\n Produto encontrado: " + str(self.itens[pos]))

    #  * imprimir todo o estoque disponível
    def printEstoque(self):
        print('\033[32m' + "\nImprimindo estoque\n".upper() + '\033[0;0m')
        for item in self.itens:
            if item != '':
                print(item)
            

    #  * excluir produto
    def excluirProduto(self):
        print('\033[32m' + "\nExcluir Item\n".upper() + '\033[0;0m')
        self.pos = ctIndice() 
        if self.pos == -1: 
            print("\nErro: Item não encontrado") 
            return 
        for i in range(0, self.num_itens):
            self.itens[i] = self.itens[i+1]
        
        self.num_itens -= 1
        print("\n Item excluido")

    #  * Vender produto (remover)
    def removerProduto(self):
        print('\033[32m' + "\nVender Item\n".upper() + '\033[0;0m')
        self.qte = 0
        self.pos = ctIndice()
        if pos == -1: 
            print("\nErro: Item não encontrado") 
            return 

        try:
            try:
                print("Item "+ str((estoque[pos]).getNome()) + "encontrado")
                print("Insira a quantidade que deseja retirar: ")
                self.qte = int(input())

                if self.qte <= 0: 
                    print("\nErro: Quantia inválida") 
                    return 

            except Exception as e: print("\nErro: Houve um problema ao ler os valores de entrada") 
            return 

            (self.itens[pos]).remover(self.qte)
            print("\n Remoção realizada com sucesso \nNovo estoque: "+ (self.itens[pos]).getself.Qte())

        except Exception as e:
            print(f"\nErro: {e}")
        

File: Lista3/Exec4/src/exec4/Loja.java
public class Loja {

    private static final int TAM = 1000;
    Estoque[] itens = new Estoque[TAM];
    int num_itens = 0;

    /**
     * cadastro de produto
     */
    public void cadastraProduto(){
        System.out.println("\nAdicionar Item \n\n");

        System.out.println("1) Inserir Livro");
        System.out.println("2) Inserir CD");
        System.out.println("3) Inserir DVD");
        System.out.println("");

        int modo = -1; 
        int qte = 0; 
        int num = -1;
        String nome = "";
        try{ 
            modo = EntradaTeclado.leInt(); 
        }
        catch (Exception e){
            System.out.println("\nErro: Bad input");
            return;
        }

        try{
            System.out.printf("Código de barras para este produto: ");
            num = EntradaTeclado.leInt();

            if(ctProduto_codigo(num) != -1){ 
                System.out.printf("Erro: produto já cadastrado\n"); 
                return; 
            }

            System.out.printf("Insira nome do produto: ");
            nome = EntradaTeclado.leString();

            if(ctProduto_codigo(nome) != -1){ 
                System.out.printf("Erro: produto já cadastrado\n"); 
                return; 
            }

            System.out.printf("Insira uma qte inicial para o estoque se quiser: ");

            try {
                qte = EntradaTeclado.leInt();
                if(qte < 0){ 
                    System.out.printf("\n>> A quantidade inicial minima é 0");
                    qte = 0;
                }
            } catch(Exception e){ 
                qte = 0; 
            }
        } catch(Exception e){ 
            System.out.println("\nErro: Não foi possível ler os valores da entrada"); 
            return; 
        }

        switch(modo){
            case 1: itens[num_itens++] = new Livro(nome, num, qte); 
                break;
            case 2: itens[num_itens++] = new CD(nome, num, qte); 
                break;
            case 3: itens[num_itens++] = new DVD(nome, num, qte); 
                break;
            default: System.out.println("\nErro: Houve um problema ao ler o modo");
        }
    }


    /**
     * Adicionar mais estoque a um item ja existente
     */
    public void adicionarProduto(){
        String flag = "";
        int qte = 0;

        System.out.println("\n Adicionar Estoque \n");
        int pos = ctIndice();
        if(pos == -1){ System.out.println("Erro \nItem não encontrado"); return; }

        try {
            System.out.println("Item "+ itens[pos].getNome() + "encontrado");
            System.out.printf("Insira a quantidade que deseja acrescentar: ");
            qte = EntradaTeclado.leInt();

            if(qte <= 0){ System.out.println("\nErro: Quantia inválida"); return; }

        } catch(Exception e){ System.out.println("\nErro: Houve um problema ao ler os valores de entrada"); return; }

        itens[pos].adicionar(qte);
        System.out.println("\n Quantia adicionada com sucesso ");
        System.out.println("Quantidade atual: "+ itens[pos].getQte());
    }


    /**
     * Encontrar produtos através do código de barras
     * @param num
     * @return
     */
    public int ctProduto_codigo(int num){
        for(int i = 0; i < num_itens; i++){
            if(num == itens[i].getNum()){ 
                return i;
            }
        }
        return -1;
    }

    /**
     * Encontrar produtos através do nome
     * @param nome
     * @return
     */
    public int ctProduto_codigo(String nome){  
        for(int i = 0; i < num_itens; i++){
            if(nome.equals(itens[i].getNome())){ 
                return i; 
            }
        }
        return -1;
    }


    /**
     * Le dados de busca do usuário e retorna a posição do produto
     * @return
     */
    public int ctIndice(){
        String flag = "";
        int qte = 0, pos = -1;

        try {
            System.out.printf("Insira o nome ou código de barras do item que deseja modificar ");
            flag = EntradaTeclado.leString();

            try {
                int num = Integer.parseInt(flag);
                pos = ctProduto_codigo(num);
                if(pos == -1){ pos = ctProduto_codigo(flag); } 
                //Faz a busca pelo nome caso nao encontre pelo codigo (nome pode ser apenas numeros)
            } 
            catch(Exception e){
                pos = ctProduto_codigo(flag);
            }
            
        } catch(Exception e){ System.out.println("\nErro: Houve um problema ao ler os valores de entrada"); return -2; }

        return pos;
    }

    
    /**
     * imprimir todo o estoque disponível
     */
    public void printEstoque(){
        System.out.println("\n Imprimindo estoque \n");
        for(int i = 0; i < num_itens; i++){
            System.out.println(itens[i]);
        }
    }


    /**
     * excluir produto
     */
    public void excluirProduto(){
        System.out.println("\n Excluir Item \n");
        int pos = ctIndice(); 
        if(pos == -1){ 
            System.out.println("\nErro: Item não encontrado"); 
            return; }
        for(int i = pos; i < num_itens; i++){ 
            itens[i] = itens[i+1];
        }
        num_itens--;
        System.out.println("\n Item excluido");
    }

    /**
     * buscar e imprimir certo produto
     */
    public void getProduto(){
        System.out.println("\n Buscar Item \n");
        int pos = ctIndice();
        if(pos == -1){ System.out.println("\nErro: Item não encontrado"); return; }
        System.out.println("\n Produto encontrado \n"+ itens[pos]);
    }

    /**
     * Vender produto (remover)
     */
    public void removeProduto(){
        System.out.println("\n Vender Item \n");
        int qte = 0, pos = ctIndice();
        if(pos == -1){ System.out.println("\nErro: Item não encontrado"); return; }

        try {
            try {
                System.out.println("Item "+ itens[pos].getNome() + "encontrado");
                System.out.printf("Insira a quantidade que deseja acrescentar: ");
                qte = EntradaTeclado.leInt();

                if(qte <= 0){ System.out.println("\nErro: Quantia inválida"); return; }

            } catch(Exception e){ System.out.println("\nErro: Houve um problema ao ler os valores de entrada"); return; }

            itens[pos].remover(qte);
            System.out.println("\n Remoção realizada com sucesso \nNovo estoque: "+ itens[pos].getQte());
        
        } catch(Exception e){
            System.out.println("\nErro: A quantia fornecida é maior que a disponível em estoque");
        }
    }

}